src/traits.rs

```diff
@@ -92,26 +98,41 @@ impl<M, O, N> Step<M, O, N> {
         self
     }
 
-    /// Converts `self` into a step of another type, given conversion methods for output and
-    /// messages.
-    pub fn map<M2, O2, FO, FM>(self, f_out: FO, f_msg: FM) -> Step<M2, O2, N>
+    /// Converts `self` into a step of another type, given conversion methods for output, faults,
+    /// and messages.
+    pub fn map<M2, O2, F2, FO, FF, FM>(
+        self,
+        f_out: FO,
+        f_fail2fail: FF,
+        f_msg: FM,
+    ) -> Step<M2, O2, N, F2>
     where
+        F2: Fail,
         FO: Fn(O) -> O2,
+        FF: Fn(Fault<N, F>) -> Fault<N, F2>,
         FM: Fn(M) -> M2,
     {
         Step {
             output: self.output.into_iter().map(f_out).collect(),
-            fault_log: self.fault_log,
+            fault_log: self.fault_log.into_iter().map(f_fail2fail).collect(),
```

Comment:
(See comment above.) This should probably convert only the fault kind, and could use the `Fault::map` method:

```rust
fault_log: self.fault_log.into_iter().map(|fault| fault.map(f_fault)).collect(),
```

(Same for `extend_with`.)